lib/streamlit/commands/query_params.py

```diff
@@ -140,3 +141,47 @@ def _ensure_no_embed_params(
         separator = "&" if current_embed_params else ""
         return separator.join([query_string, current_embed_params])
     return current_embed_params
+
+
+@dataclass
+class QueryParams:
+    """
+    A dict-like representation of query params that generally behaves like st.session.state.
+    The main difference is that it only stores and returns str and list[str].
+
+    TODO(willhuang1997): Fill in these docs with examples and fix doc above. Above is just a stub for now.
+    """
+
+    query_params: Dict[str, List[Any]] = field(default_factory=dict)
+
+    def __init__(self, query_params: Dict[str, List[Any]] = {}):
+        self.query_params = query_params
+
+    def __repr__(self):
+        return util.repr_(self)
+
+    def __contains__(self, key: str) -> bool:
+        return key in self.query_params
+
+    def __len__(self):
+        return len(self.query_params)
+
+    def _send_query_param_msg(self):
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return
+        msg = ForwardMsg()
+        msg.page_info_changed.query_string = _ensure_no_embed_params(
+            self.query_params, ctx.query_string
+        )
+        ctx.query_string = msg.page_info_changed.query_string
+        ctx.enqueue(msg)
+
+    def clear(self):
+        self.query_params.clear()
+        self._send_query_param_msg()
+
+    def __delitem__(self, key: str) -> None:
+        if key in self.query_params:
+            del self.query_params[key]
```

Comment:
ditto here:
I think it's ok to do this as if the ctx is None:, they're running it with some python code and I don't think we really care about that case.